cmd/goal/clerk.go

```diff
@@ -724,46 +724,77 @@ var signCmd = &cobra.Command{
 
 		var outData []byte
 		dec := protocol.NewDecoderBytes(data)
+		// read the entire file and prepare in-memory copy of each signed transaction, with grouping.
+		txnGroups := make(map[crypto.Digest][]*transactions.SignedTxn)
+		groupsOrder := make([]crypto.Digest, 0)
+		txnIndex := make(map[*transactions.SignedTxn]int)
 		count := 0
 		for {
-			// transaction file comes in as a SignedTxn with no signature
-			var uncheckedTxn transactions.SignedTxn
-			err = dec.Decode(&uncheckedTxn)
+			uncheckedTxn := new(transactions.SignedTxn)
+			err = dec.Decode(uncheckedTxn)
 			if err == io.EOF {
 				break
 			}
 			if err != nil {
 				reportErrorf(txDecodeError, txFilename, err)
 			}
+			group := uncheckedTxn.Txn.Group
+			if group.IsZero() {
+				// create a dummy group.
+				randGroupBytes := crypto.Digest{}
+				crypto.RandBytes(randGroupBytes[:])
+				group = randGroupBytes
+			}
```

Comment:
I don't think there's anything wrong here, but I would like to understand how this works.  I expected that the rule for group id was that if it was set, it must be the hash of the entire group, but this code makes me think there's nothing in the blockchain that checks that (since you're allocating a random group id).  Is that right?  The "rule" that a group id must be the hash of the whole group is just a best practice that should be done by users as they decide what transactions to sign. (If they check the hash, they know they're not signing something surprising.) But the blockchain does not care.

Reply:
The blockchain does care. look in `func (eval *BlockEvaluator) TestTransactionGroup(txgroup []transactions.SignedTxn) error`.

The group here is used *only* for the purpose of grouping the transactions from the input file before being processed. It does not *modify* the Group value within the transaction itself.
The allocation of the random group here is used only for the purpose of preserving the original transaction ordering of the input file.